### PR TITLE
`stat_smooth()` drops failed groups

### LINKED ISSUE #5352: `geom_smooth()` fails for every group if only one group is wrong
It seems that `geom_smooth()` will not draw any smooth line even if the computation fails for just one of the many groups. 

This doesn't work:

``` r
library(ggplot2)
packageVersion("ggplot2")
#> [1] '3.4.2.9000'

diamonds |> 
  subset(cut == "Ideal" & color == "J") |>
  ggplot( aes(carat, price)) +
  geom_point(aes(color = clarity)) +
  geom_smooth(aes(color = clarity))
#> `geom_smooth()` using method = 'loess' and formula = 'y ~ x'
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : span too small.  fewer data values than degrees of freedom.
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : at 0.9598
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : radius 0.00010404
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : all data on boundary of neighborhood. make span bigger
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : pseudoinverse used at 0.9598
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : neighborhood radius 0.0102
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : reciprocal condition number 1
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : at 3.0202
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : radius 0.00010404
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : all data on boundary of neighborhood. make span bigger
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : There are other near singularities as well. 0.00010404
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : zero-width neighborhood. make span bigger

#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : zero-width neighborhood. make span bigger
#> Warning: Computation failed in `stat_smooth()`
#> Caused by error in `predLoess()`:
#> ! NA/NaN/Inf in foreign function call (arg 5)
```

The issue is that the "I1" clarity group has just two observations, and therefore `geom_smooth()` can't fit anything. If I remove those data, now it works: 

``` r
diamonds |> 
  subset(cut == "Ideal" & color == "J" & clarity != "I1") |> 
  ggplot( aes(carat, price)) +
  geom_point(aes(color = clarity)) +
  geom_smooth(aes(color = clarity))
#> `geom_smooth()` using method = 'loess' and formula = 'y ~ x'
```

This is surprising because I would've thought that each group is fitter independently and even if one group fails to fit, the rest should work. 

<sup>Created on 2023-07-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #5352.

Briefly, it wraps the fitting/predicting procedure in a `tryCatch()` block to rethrow errors as warnings and drops the group for which the procedure fails. This prevents one bad group stopping the fitting in all other groups.

The example from #5352 with this PR (in particular, see the very last warning):

``` r
devtools::load_all("~/packages/ggplot2")
#> ℹ Loading ggplot2
#> Warning: package 'testthat' was built under R version 4.3.1

diamonds |> 
  subset(cut == "Ideal" & color == "J") |>
  ggplot( aes(carat, price)) +
  geom_point(aes(color = clarity)) +
  geom_smooth(aes(color = clarity))
#> `geom_smooth()` using method = 'loess' and formula = 'y ~ x'
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : span too small.  fewer data values than degrees of freedom.
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : at 0.9598
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : radius 0.00010404
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : all data on boundary of neighborhood. make span bigger
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : pseudoinverse used at 0.9598
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : neighborhood radius 0.0102
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : reciprocal condition number 1
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : at 3.0202
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : radius 0.00010404
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : all data on boundary of neighborhood. make span bigger
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : There are other near singularities as well. 0.00010404
#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : zero-width neighborhood. make span bigger

#> Warning in simpleLoess(y, x, w, span, degree = degree, parametric = parametric,
#> : zero-width neighborhood. make span bigger
#> Warning: Failed to fit group 1.
#> Caused by error in `predLoess()`:
#> ! NA/NaN/Inf in foreign function call (arg 5)
```

<sup>Created on 2023-07-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/stat-smooth.R [hunk 1]
```diff
@@ -171,14 +171,25 @@ StatSmooth <- ggproto("StatSmooth", Stat,
...
+      },
+      error = function(cnd) {
+        cli::cli_warn("Failed to fit group {data$group[1]}.", parent = cnd)
+        NULL
+      }
+    )
+    if (is.null(prediction)) {
+      return(NULL)
+    }
+
```
Comment (thomasp85) [inline R/stat-smooth.R:192]: We prefer to use the slang error handling when possible (`try_fetch()`)

**Review by thomasp85 — CHANGES_REQUESTED** [review]
I'm not sure what we can do about the excessive warnings. Summarising them would require us to make some very hard expectations about the warnings coming from the call...

File: R/stat-smooth.R (hunk 1 above)
Comment (teunbrand) [reply R/stat-smooth.R]: Thanks for pointing this out. I'm not up to speed with what the difference exactly is, but that seems to work well here too.

**Review by thomasp85 — APPROVED** [review]
LGTM

**Pull request merged**